xds/src/test/java/io/grpc/xds/XdsNameResolverTest.java

```diff
@@ -102,6 +102,74 @@ public void invalidName_hostnameContainsUnderscore() {
     }
   }
 
+  /**
+   * Each resolver explicitly triggers reading bootstrap configurations.
+   */
+  @Test
+  public void eachResolverReadsBootstrapSeparately() {
+    final ChannelCreds loasCreds = new ChannelCreds("loas2", null);
+    final ChannelCreds googleDefaultCreds = new ChannelCreds("google_default", null);
+    Bootstrapper bootstrapper1 = new Bootstrapper() {
+      @Override
+      BootstrapInfo readBootstrap() throws Exception {
+        return new BootstrapInfo("trafficdirector-foo.googleapis.com",
+            ImmutableList.of(googleDefaultCreds), FAKE_BOOTSTRAP_NODE);
+      }
+    };
+    Bootstrapper bootstrapper2 = new Bootstrapper() {
+      @Override
+      BootstrapInfo readBootstrap() throws Exception {
+        return new BootstrapInfo("trafficdirector-bar.googleapis.com",
+            ImmutableList.of(loasCreds), FAKE_BOOTSTRAP_NODE);
+      }
+    };
+
+    XdsNameResolver resolver1 = new XdsNameResolver("service-foo.googleapis.com", bootstrapper1);
+    XdsNameResolver resolver2 = new XdsNameResolver("service-bar.googleapis.com", bootstrapper2);
```

Comment:
This is not the right test. You provide different bootstrappers, then of course the results are different. The real XdsNameResolver is always using the same bootstrapper, the default instance. Actually in this situation there is nothing that is nice to have a test. If you do want to add a test, you provide only one bootstrapper, but `readBootstrap()` returns different values at different times when you call it. Yes, of course the results are different (only to verify it's not caching the first result), so anyway, I don't think it really worth a test.

Reply:
I am ok with either changing it (provide only one bootstrapper) or deleting it.

Reply:
Deleted. Even using a the same `Bootstrapper` instance (stubbed to return different results for different invocations) doesn't seem right and doesn't really test anything. I would not bother having such a test.

Reply:
The real test would be to instantiate a `Bootstrapper` and calls `readBootstrap` multiple times to show it returns different results (based on current contents). I think that's possible and worth the effort (you never know when a test will find regressions). However I am going to LGTM to unblock this PR

Reply:
All the confusion comes from the specification of `Bootstrapper.readBootstrap()` method (not explicitly written out in head comment but inferred from implementation). We cannot tell if this method reads bootstrap explicitly every time or just returns a cached copy after first read. This is `Bootstrapper`'s implementation detail and callers doesn't even need to know. Callers only need to know calling `readBootstrap()` returns a `BootstrapInfo` that contains data they need. Therefore, a test that verifies `readBootstrap()` method is called multiple times does really mean anything.

However, if we do want to verify that we are reading file explicitly every time. That should be a test for `readBootstrap()`'s implementation. We don't have such a test. Ideally we would need to add such a test. However, this is something "easy to implement but hard to verify". The implementation shows that we are reading a file from local system every time the method executes very obviously. That is no doubt for this by just taking a glance of the code. However, verifying this in a unit test needs resources setup such as using `org.junit.rules.TemporaryFolder`. We would not bother do sophisticated testing for obvious implementation.

Anyway, thanks for pointing this out and sorry for making this confusion.

Reply:
I realize the point that you can look at the code and convince yourself of how/what it does. However the very title and purpose of this PR is "do not globally cache results of reading bootstrap file" because of an issue that was pointed out. Consider the situation that someone in the future looks at the code and decides to "optimize" it by caching `BootstrapInfo" and breaks it unknowingly. A test could easily prevent that.

Reply:
Nah... The title should be "do not cache results of reading bootstrap file". But, well, too later...